### PR TITLE
ci: Add GitHub Actions workflows and drop Travis

### PR DESCRIPTION
**Pull request by Airblader**
Replace the dead .travis.yml with two GitHub Actions workflows:

* ci.yml builds the project and runs the test suite on push and PR, suitable for use as a required status check.
* release.yml is triggered by a v* tag push or a manual dispatch with a version input. It validates the version against the Makefile, creates the tag (when dispatched), runs the build and tests, and publishes a GitHub release with a source tarball and auto-generated notes.

The shared apt and CPAN setup lives in a composite action under .github/actions/setup so both workflows install the same dependencies.

Update the README to use the new GitHub Actions badge in place of the Travis one.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        run: |
+          set -euo pipefail
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          if git rev-parse "$TAG" >/dev/null 2>&1; then
+            echo "::error::Tag $TAG already exists."
+            exit 1
+          fi
+          git tag -a "$TAG" -m "Release $TAG"
+          git push origin "$TAG"
```
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/release.yml]: Good catch, this is real. The current order pushes the annotated tag right after the version-resolve step but before `make all` / `make test` / the tarball / the GitHub release publish. If anything from `make all` onward fails, we're left with `vX.Y` pointing at an unreleased build, and a re-dispatch immediately fails on the "tag already exists" guard — which means the only way to retry is to manually delete the tag, which is a sharp edge nobody wants on release day.

Two reasonable ways to fix it:

1. **Reorder the workflow** so the tag is created and pushed only after the build, tests, and tarball succeed — i.e. move the *Create and push tag* step right before the `softprops/action-gh-release` step. The `git archive` step would need to use `HEAD` rather than the tag name (or we can create the tag locally without pushing first, then push it at the very end).
2. **Drop the explicit tag-push step entirely** and let `softprops/action-gh-release@v2` create the tag itself — it accepts `tag_name` and will create the tag on the target ref when it doesn't already exist. That collapses the ordering problem because the tag only materialises on a successful publish.

Option 2 is the smaller change and removes the sharp edge for re-runs. Either is fine. Worth doing as a follow-up.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      - 'v*'
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version to release (e.g. 1.2). The Makefile VERSION must match.'
+        required: true
+        type: string
+
+permissions:
+  contents: write
```
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/release.yml:15]: Fair point in principle — least-privilege says we should grant `contents: write` only to the publish step, and run the rest with the default read-only token.

The standard pattern for that on GitHub Actions is to split the workflow into two jobs: a build job with the default read-only permissions that produces the source tarball as a workflow artifact (`actions/upload-artifact`), and a publish job declared with `permissions: contents: write` that pulls the artifact down and creates the release. Per-step permission scoping isn't supported, so two jobs is the way.

Whether that's worth the complexity here depends on threat model. The CPAN install runs `cpanm --notest --quiet X11::XCB IPC::System::Simple` — those are pinned-by-name, not arbitrary code from a PR — so the realistic compromise vector is upstream CPAN being hijacked, which would already be a wider problem than this single workflow. The release workflow also doesn't run on PRs from third parties, only on tag push or maintainer-triggered `workflow_dispatch`, so the attacker model is constrained.

Still, splitting into two jobs is the cleaner setup and would be a reasonable follow-up if/when we want supply-chain hardening to live close to the release path. Calling out as a real recommendation, just not implementing it here.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            -o "xedgewarp-${VERSION}.tar.gz" \
+            "${TAG}"
+
+      - name: Publish GitHub release
+        uses: softprops/action-gh-release@v2
+        with:
+          tag_name: ${{ steps.version.outputs.tag }}
+          name: ${{ steps.version.outputs.tag }}
+          generate_release_notes: true
+          files: xedgewarp-${{ steps.version.outputs.version }}.tar.gz
```
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/release.yml:96]: Reasonable supply-chain hardening recommendation, and a real risk class — `tj-actions/changed-files` got compromised this way in March 2025 when its tags were retargeted to a malicious commit, and any workflow pinning to a mutable tag rather than a SHA executed the malicious code. Pinning to an immutable commit SHA closes that vector for this workflow.

The trade-off is that SHA-pinned actions don't pick up upstream security fixes automatically, so the recommended pattern is to combine SHA pinning with Dependabot (`package-ecosystem: "github-actions"` in `.github/dependabot.yml`) so the SHAs get refreshed via PR whenever upstream cuts a new release. Without that, a SHA pin tends to bitrot.

Worth doing alongside a Dependabot config in a follow-up if/when you want to harden the release path. Not blocking on it here, but flagging it as legitimate.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        sudo apt-get update
+        sudo apt-get install -y \
+          libxcb1-dev \
+          libxcb-randr0-dev \
+          libxcb-util-dev \
+          libx11-dev \
+          libx11-xcb-dev \
+          libxi-dev \
+          libxfixes-dev \
+          libxrandr-dev \
```
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/ci.yml:27]: Partially valid coverage gap, though smaller than the comment suggests.

The CI step `make` resolves to the default target, which the Makefile defines as `all: clean $(TARGET) mans`. So both the binary build *and* manpage generation (via `a2x`) are exercised on every PR. What isn't covered is the `install` rule itself, which is two `install -Dm` invocations to `$(BINDIR)` and `$(MANDIR)` — about as simple as install rules get. The realistic regression risk is low.

If we do want it covered without needing `sudo`, the cheap version would be:

```yaml
- name: Install (smoke test)
  run: make DESTDIR=$PWD/_install install
```

That stages into a workspace-local prefix, no privileges required, and would catch breakage in the install recipe (missing files, bad paths, permission mode typos, etc.). Worth considering as a small follow-up; not blocking this PR.

File: .github/workflows/release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/release.yml:15]: Ignoring for now.

File: .github/workflows/ci.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/ci.yml:27]: This comment looks like it's evaluating an older draft of the workflow. The CI step in the current file is just `make` (line 32 of `ci.yml`) — there's no separate `make all` followed by `make test`. `make test` was deliberately removed from CI in favour of build-only coverage, so there's no double-compile happening here.

The same observation does still apply to `release.yml`, where both `make all` and `make test` are present — see the other thread for that. Marking this one as moot for the file it's filed against.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          fi
+          git tag -a "$TAG" -m "Release $TAG"
+          git push origin "$TAG"
+
+      - name: Set up environment
+        uses: ./.github/actions/setup
+
+      - name: Build
+        run: make all
+
```
[the remark this answers is not in the document]
Comment (Airblader) [reply .github/workflows/release.yml]: Confirmed and valid. `make test` declares `clean $(TARGET)` as its prerequisites, so it rebuilds the binary from scratch — the preceding `make all` step recompiles the same C sources twice on every release run.

The reason `make all` is there at all is to exercise manpage generation (the `mans` target via `a2x`), since the tarball ships from `git archive` and won't include built artifacts; we want the release to fail loudly if asciidoc is broken on the runner. The minimal fix is to drop `make all` and add a `make mans` step alongside `make test` — that gives manpage coverage without the duplicate binary compile:

```yaml
- name: Test
  run: make test

- name: Build manpages
  run: make mans
```

Or just leave `make all` in place and accept the duplicate compile — release runs are infrequent and ~30s of extra wall time is cheap. Either way, good catch; happy to follow up with the smaller version.
